### ISSUE TITLE
object 'kenpom' not found

### ISSUE THREAD

**Issue by aqsmith08** [sportsdataverse/hoopR#94]
**Describe the bug**
It seems like something is broken with KenPom functions in hoopR (or I'm doing something wrong). I'm also noticing that documentation on the hoopR [website](https://hoopr.sportsdataverse.org/reference/kp_hca.html?q=object%20kenpom#ref-examples) is displaying the error too. Screenshot below.

**To Reproduce**
```
> library(hoopR)
> hoopR::kp_hca()
Error in hoopR::kp_hca() : object 'kenpom' not found
> hoopR::kp_conf(2021, conf = "B10")
Error in hoopR::kp_conf(2021, conf = "B10") : object 'kenpom' not found

```
**Expected behavior**
It does look like values do exist -  https://kenpom.com/hca.php - and I would expect that those values would be returned.

**Screenshots**
![image](https://user-images.githubusercontent.com/5125743/198118502-4a76f410-6333-4f1c-a16a-caaa4a3cab96.png)

![image](https://user-images.githubusercontent.com/5125743/198119063-0d63190f-efdb-4b45-ab2d-7f7493f183b6.png)

**Comment by aqsmith08** [comment]
I checked my username & password and that seems to be just fine.
```
> has_kp_user_and_pw()
[1] TRUE
```
When I went into the `login` function specifically and ran it line by line, I hit an error here:
```
>       browser <- login()
Error in read_html.response(x$response, ..., base_url = x$url) : 
  Forbidden (HTTP 403).
In addition: Warning message:
In session_set_response(x, resp) : Forbidden (HTTP 403).
```
Could kenpom.com be blocking the traffic?

**Comment by aqsmith08** [comment]
Looks like there was a recent [StackOverflow post](https://stackoverflow.com/questions/74127907/error-in-read-html-responsexresponse-base-url-xurl-forbidden-http) about the same error. I may try to reset my kenpom.com PW and see if it works.

**Comment by aqsmith08** [comment]
Password reset and updating environment via `usethis::edit_r_environ()` didn't seem to resolve the issue.
```
> login(user_email = Sys.getenv("KP_USER"), user_pw = Sys.getenv("KP_PW"))
Error in read_html.response(x$response, ..., base_url = x$url) : 
  Forbidden (HTTP 403).
In addition: Warning message:
In session_set_response(x, resp) : Forbidden (HTTP 403).
```

**Comment by saiemgilani** [comment]
Yeah, can you confirm you're on v1.8.1+? It's definitely something I'm not really figuring out why it is happening on the docs website still but many seem to resolve the issue locally with the update to 1.8.1. Keep you posted

**Comment by aqsmith08** [comment]
It looks like I might not be on the most updated version although I don't see any package updates available via RStudio. 

```
> packageVersion("hoopR")
[1] ‘1.8.0’
```
Want me to try the package version available via Github from the instructions on the README?
```
# You can install using the pacman package using the following code:
if (!requireNamespace('pacman', quietly = TRUE)){
  install.packages('pacman')
}
pacman::p_load_current_gh("sportsdataverse/hoopR", dependencies = TRUE, update = TRUE)
```

**Comment by aqsmith08** [comment]
Updating the package fixed it.
```
> packageVersion("hoopR")
[1] ‘1.8.1’
> library(hoopR)
> foo <- kp_hca()
> head(foo)
           team conf hca hca_rk   pf pf_rk  pts pts_rk  nst nst_rk blk blk_rk elev elev_rk
1    Texas Tech  B12 4.8      1 -5.5     3 11.2      7 -1.6     40 1.6     66 3200      20
2 West Virginia  B12 4.7      2 -5.2     6 12.0      2 -2.6      2 2.1     20 1000      60
3     Air Force  MWC 4.6      3 -3.8    51 10.6     13 -1.1    119 1.4    102 7000       2
4           TCU  B12 4.4      4 -5.4     4  9.8     26 -1.4     74 2.4      8  600     140
5      Colorado  P12 4.3      5 -3.6    61 10.2     20 -0.8    189 1.4    120 5400       5
6    New Mexico  MWC 4.2      6 -3.1    94 11.2      5 -0.7    199 1.1    195 5100       7
```
I'm going to close this bug and hopefully this is helpful to other folks who have this issue. No idea what is different between the versions but it's something important. Thanks for getting me heading in the right direction.

**Comment by aqsmith08** [comment]
Ok, I'm the dummy who's experiencing this again and I can't figure out how to resolve it. Here's how I recreate it:

```
> library(hoopR)
> has_kp_user_and_pw()
[1] TRUE
> kp_pomeroy_ratings(min_year = 2020, max_year = 2021)
Error in kp_pomeroy_ratings(min_year = 2020, max_year = 2021) : 
  object 'kenpom' not found
> packageVersion("hoopR")
[1] ‘2.1.0’
> 
```
When I went into the function itself and tried to run it, I received the following error:

```
<code ran>
+         ### Store Data
+         if (year == min_year) {
+           kenpom <- x
+         } else {
+           kenpom <- dplyr::bind_rows(kenpom, x)
+         }
+       }
Error in `rvest::session_jump_to()`:
! `x` must be produced by `session()`.
Run `rlang::last_trace()` to see where the error occurred.
```

Can you make anything from that? Can it be something around that function in rvest?

https://github.com/sportsdataverse/hoopR/blob/8ee009ec1a0fb16fb3d3b6ffbe0bdfc2ff051412/R/kp_ratings_tables.R#L100

My past self fixed it by updating the package which I've already done and it seems to still be an issue. Anyone have any suggestions? Thanks!

**Comment by aqsmith08** [comment]
It seems like I'm hitting a 403 error but I can't figure out if it's my incorrect permissions (which I don't think it is) or if it's anti-scraping stuff.

```
> year <- 2020
> url <- paste0("https://kenpom.com/index.php?y=", year)
> browser <- rvest::session(url)
Warning message:
In session_set_response(x, resp) : Forbidden (HTTP 403).
```

**Comment by aqsmith08** [comment]
Ok, I've poked some more and I'm now looking at the `login()` function. When I pulled that apart:

```
> user_email <- Sys.getenv("KP_USER")
> user_pw <- Sys.getenv("KP_PW")){
> url <- "https://kenpom.com/index.php"
> my_session <- rvest::session(url, httr::add_headers(.headers = .kp_headers()))
Error in .kp_headers() : could not find function ".kp_headers"
```

This drops me into the `.kp_headers` function - https://github.com/sportsdataverse/hoopR/blob/8ee009ec1a0fb16fb3d3b6ffbe0bdfc2ff051412/R/utils_nba_stats.R#L87 but I'm not sure how or what to investigate further. Could it simply be that the `utils_nba_stats.R` file isn't being sourced or something? I'm not sure I understand.

Ok, that's all I've got tonight. Any help or insight would be great. I'm happy to poke at this further.

**Comment by aqsmith08** [comment]
Per https://kenpom.com/terms.php which states (in part) --

I. You may not, except with our express written permission, use any automated means to access or use the Web Site. This includes scripts, bots, scrapers, data miners, or similar software, in a manner that adversely impacts site performance or access.

J. You may not use any material or content from the Web Site, including without limitation any statistics or data, (i) to create any database, archive, or other data store that competes with or constitutes a material substitute for the services or data stores offered on the Web Site or (ii) to provide any service that competes with or constitutes a material substitute for the services or data stores offered on the Web Site.

---
I'm pretty convinced that kenpom is blocking the package from working and that we should look elsewhere.

**Comment by saiemgilani** [comment]
Same class as #168: the 403-in-rvest-session login failure masked as `object 'kenpom' not found` was fixed by the httr2 login rewrite (c0f6ba4a6) + loud error reporting. Live-verified working on main today. Ships in 3.1.0. Closing.
